compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2542,9 +2544,30 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 self.is_slice_or_array_or_vector(resolved_ty);
             match resolved_ty.kind() {
                 ty::Adt(adt_def, _)
-                    if self.tcx.is_diagnostic_item(sym::Option, adt_def.did())
-                        || self.tcx.is_diagnostic_item(sym::Result, adt_def.did()) =>
-                {
+                    if (self.tcx.is_diagnostic_item(sym::Option, adt_def.did())
```

Comment:
Instead of doing all this, could you perhaps just check that the `pat_info.top_info.expected` is the same type as the `expected_ty`?

I'm not sure if that works or if it needs some tweaking or investigations, but basically I'd prefer that we check something along the lines of "have we begun to peel off any of the pattern types" rather than doing all this manual checking for `Result::Err` which seems hyperspecific.

Reply:
@compiler-errors 
> we check something along the lines of "have we begun to peel off any of the pattern types" rather than doing all this manual checking for Result::Err which seems hyperspecific.

 I understand. I'll fix it.

`pat_info.top_info.expected` (`.kind()` is `Infer(InferTy::TyVar)`)  is resolved to `Result<i32, Vec<ErrorKind>`,  and `expected_ty` is `Vec<ErrorKind>` (will never be the same?), so it seems necessary to do something more for checking an equality. 

However, I'm not sure if it is doing too much for the `as_deref` suggestion...